clippy_lints/src/types/redundant_allocation.rs

```diff
@@ -9,74 +9,61 @@ use rustc_span::symbol::sym;
 use super::{utils, REDUNDANT_ALLOCATION};
 
 pub(super) fn check(cx: &LateContext<'_>, hir_ty: &hir::Ty<'_>, qpath: &QPath<'_>, def_id: DefId) -> bool {
-    if Some(def_id) == cx.tcx.lang_items().owned_box() {
-        if let Some(span) = utils::match_borrows_parameter(cx, qpath) {
-            let mut applicability = Applicability::MachineApplicable;
-            span_lint_and_sugg(
-                cx,
-                REDUNDANT_ALLOCATION,
-                hir_ty.span,
-                "usage of `Box<&T>`",
-                "try",
-                snippet_with_applicability(cx, span, "..", &mut applicability).to_string(),
-                applicability,
-            );
-            return true;
-        }
+    let outer_sym = if Some(def_id) == cx.tcx.lang_items().owned_box() {
+        "Box"
+    } else if cx.tcx.is_diagnostic_item(sym::Rc, def_id) {
+        "Rc"
+    } else if cx.tcx.is_diagnostic_item(sym::Arc, def_id) {
+        "Arc"
+    } else {
+        return false;
+    };
+
+    if let Some(span) = utils::match_borrows_parameter(cx, qpath) {
+        let mut applicability = Applicability::MachineApplicable;
```

Comment:
Hey @flip1995, I've tested the current implementation of this lint on this *very useful* function [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8e56aa3623759c811660250cab26b243):
```rs
#[allow(dead_code)]
pub fn test_fn(useful_value: &u128) -> Box<&u128> {
    Box::new(useful_value)
}
```
The suggested refactoring looks like this:
```rs
pub fn test_fn(useful_value: &u128) -> &u128 {
    Box::new(useful_value)
}
```
Which is invalid due to a type mismatch. I believe that this is a problem with every lint that suggest changes to types, is there any special rule that we allow applicability on types even if that could cause invalid code? Otherwise, we should change it to `Unspecified` IMO :thinking:. (The previous implementation also used `Applicability::MachineApplicable`)

Sorry, for pinging you so much lately :sweat_smile: 

[`Applicability` documentation](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/enum.Applicability.html)

Reply:
We usually use `MaybeIncorrect` in those cases, even though it isn't meant for that use case. But there isn't anything better that says "This suggestion is probably wrong, since it requires further changes".

> Sorry, for pinging you so much lately

No problem, keep doing it :)

---

Sidenote: Since this lint changes function signatures, it should probably not apply to public items by default. We have a config flag for this.

Reply:
May I change `MachineApplicable` to `MaybeIncorrect` in `redundant_allocation`?

Reply:
> Sidenote: Since this lint changes function signatures, it should probably not apply to public items by default. We have a config flag for this.

You're probably referring to [`avoid_breaking_exported_api`](https://github.com/rust-lang/rust-clippy/blob/2464ee92840f8d081bb2f546020581e73ab789fa/clippy_lints/src/utils/conf.rs#L125-L126). Supporting that configuration on all type lints would be a good step, but it would require changes to the type module in general. I think it would be better to do that with other type lints together in a different PR. What do you think? @flip1995 I could create an issue for it. :upside_down_face:

Reply:
Hey @lengyijun, could you change the applicability to `MaybeIncorrect` as discussed earlier.

Reply:
It isn't necessary for all type lints. Just for type lints that may suggest changing the signature of public API, for example function signatures. So, I wouldn't add it to all type lints, just think that it might make sense for this lint, just from the example, you gave above.